concepts/websockets.textile

```diff
@@ -0,0 +1,591 @@
+
+*Server (Node.js):*
+
+```[js]
+var WebSocketServer = require('uws').Server;
+var wss = new WebSocketServer({ port: 3000 });
+ 
+function onMessage(message) {
+  console.log('received: ' + message);
+}
+ 
+wss.on('connection', function(ws) {
+  ws.on('message', onMessage);
+  ws.send('something');
+});
+```
+
+h3(#faye-websocket). faye-websocket
+
+"faye-websocket":https://github.com/faye/faye-websocket-node is a standards-compliant WebSocket implementation for both the client and server, and originated from the Ruby-on-Rails community as part of the "Faye project":https://faye.jcoglan.com/. As per the Github, project:
+
+bq. It does not provide a server itself, but rather makes it easy to handle WebSocket connections within an existing Node application. It does not provide any abstraction other than the standard WebSocket API.
+
+
+
+In the sample code below for the server, you can see that all of the work of handling the connection upgrade and translating messaging frames from inbound socket buffers is handled by the @WebSocket@ class provided by the library. As with other minimal solutions, this is a no-frills implementation - you'll need to handle application-specific concerns yourself.
+
+h4(#minimal-code-samples-1). Minimal code samples
+
+*Client (Browser, before bundling):*
+
+```[js]
+var WebSocket = require('faye-websocket'),
+    ws        = new WebSocket.Client('ws://www.example.com/');
+
+ws.on('open', function(event) {
+  console.log('open');
+  ws.send('Hello, world!');
+});
+
+ws.on('message', function(event) {
+  console.log('message', event.data);
+});
+
+ws.on('close', function(event) {
+  console.log('close', event.code, event.reason);
+  ws = null;
+});
+```
+
+*Server (Node.js):*
+
+```[js]
+var WebSocket = require('faye-websocket'),
+    http      = require('http');
+
+var server = http.createServer();
+
+server.on('upgrade', function(request, socket, body) {
+  if (WebSocket.isWebSocket(request)) {
+    var ws = new WebSocket(request, socket, body);
+    
+    ws.on('message', function(event) {
+      ws.send(event.data);
+    });
+    
+    ws.on('close', function(event) {
+      console.log('close', event.code, event.reason);
+      ws = null;
+    });
+  }
+});
+
+server.listen(8000);
+```
+
+h3(#socket.io). Socket.io
+
+"Socket.io":https://github.com/socketio/socket.io has been around for a while now; I like to think of it as the "jQuery" of WebSockets. It uses long polling and WebSockets for its transports, by default starting with long polling and then upgrading to WebSockets if available. Given the waning relevance of long polling, Socket.io's main drawcards these days are its "other features":https://stackoverflow.com/questions/38546496/moving-from-socket-io-to-raw-websockets/38546537#38546537, such as restoring dropped connections, automatic support for JSON, and "namespaces", which are essentially isolated messaging channels multiplexed over the same client connection.
```

Comment:
Avoid using first person references, please. 'for a while now; think of it as the "jQuery" of WebSockets.' would be fine.

Reply:
Fixed.